### ISSUE TITLE
How to keep service worker from intercepting WP REST requests?

### ISSUE THREAD

**Issue by reggieofarrell** [xwp/pwa-wp#1106]
We are using this plugin on showlists.net. It's a concert listing site. I recently ran into an issue where the service worker is intercepting WP REST requests and returning a 200 with what I assume is the 500 template. This was a custom WP REST endpoint of ours that a bug was introduced into. So the endpoint was actually returning a 500, but the user would have no idea that the request failed because the service worker is returning a 200 along with the 500 template HTML. That is having the effect of our error messaging not being shown because the request is getting back a 200 when it should be getting back a 500.

If I inspect the response from the service worker in Chrome, I see this in the Preview tab...

<img width="866" alt="Screen Shot 2024-02-09 at 6 32 53 PM" src="https://github.com/GoogleChromeLabs/pwa-wp/assets/18563006/13f40a2e-6e93-4a9e-8bc1-f26f89d6cb42">

Is there a way to configure the service worker to ignore calls to any routes that start with `/wp-json`?

**Comment by westonruter** [comment]
This should be the filter you're looking for:

https://github.com/GoogleChromeLabs/pwa-wp/blob/b4e13c7aa77cc45fbfbcd694d4e37d2500bb6cf5/wp-includes/components/class-wp-service-worker-navigation-routing-component.php#L408-L419

Using plugin code like this:

```php
add_filter(
	'wp_service_worker_navigation_route_denylist_patterns',
	static function ( $patterns ) {
		$patterns[] = '^\/wp-json';
		return $patterns;
	}
);
```

**Comment by reggieofarrell** [comment]
@westonruter Thanks. I gave that a try and the initial request was still intercepted by the service worker which returned a 200 even though the endpoint returned a 500...

<img width="862" alt="Screen Shot 2024-02-27 at 5 56 23 PM" src="https://github.com/GoogleChromeLabs/pwa-wp/assets/18563006/e3e09044-5092-4b92-a34b-e020aa58430b">

<img width="857" alt="Screen Shot 2024-02-27 at 5 56 29 PM" src="https://github.com/GoogleChromeLabs/pwa-wp/assets/18563006/0138394b-28a7-4d61-ab17-87e08e6745f7">

I added the filter in a plugin where we have all of our custom back-end code. Does that filter need to be wrapped in another action potentially?

**Comment by westonruter** [comment]
No, it doesn't matter where you add the filter. 

I see it showing up in [your service worker](https://austin.showlists.dev/wp.serviceworker):

```js
	const denylist = ["^\\\/wp-json","^\\\/wp\\-admin($|\\?|\/)","^[^\\?]*?\\.php($|\\?)","\\?(.*?&)?wp_service_worker=","^[^\\?]*?\\\/wp\\.serviceworker(\\?|$)","^[^\\?]*?\\\/feed\\\/(\\w+\\\/)?$","\\?(.*?&)?wp_customize=","\\?(.*?&)?customize_changeset_uuid=","^\\\/wp\\-json\\\/"].map(
		(pattern) => new RegExp(pattern)
	);
```

Actually I even see `wp-json` appearing twice. This is because I forgot that the REST API was already supposed to be getting excluded from handling navigation requests:

https://github.com/GoogleChromeLabs/pwa-wp/blob/6134874169e87af7df801d139e727a4950111524/wp-includes/components/class-wp-service-worker-navigation-routing-component.php#L439-L440

So no need to filter `wp_service_worker_navigation_route_denylist_patterns` as this is redundant.

How can I reproduce the issue you're seeing on https://austin.showlists.dev/ ?

**Comment by reggieofarrell** [comment]
@westonruter  Yeah, I just looked at the code and noticed that line that was supposed to be excluding the REST api already as well. Hmmm. I can add you to our repo if you want to look at the code. I can't even recreate it locally because Chrome doesn't like the local SSL cert that was created by Local WP. So, the service worker doesn't load locally. I have to push to this staging site (.dev) in order to test anything PWA related which makes it kind of slow to diagnose. Right now I have the show submission endpoint on the `.dev` site just returning an error every time for the purpose of testing this.

**Comment by reggieofarrell** [comment]
@westonruter to answer your question on how to reproduce...

1. Create a custom REST endpoint that returns a 500
2. Call it from the browser

You should see that the initial call get's intercepted by the service worker which then returns a 200 instead of the 500 that it should be returning.

**Comment by westonruter** [comment]
> 2\. Call it from the browser

Call it in what way? By navigating to it or by requesting it with `fetch()`?

**Comment by reggieofarrell** [comment]
@westonruter Sorry. Yes, requesting it with `fetch()` from a page where the service worker is loaded.

**Comment by westonruter** [comment]
OK, here is a plugin with which I can reproduce the issue:

```php
<?php
/**
 * Plugin Name: Try 500 REST API error
 */

add_action( 'rest_api_init', static function () {
	register_rest_route(
		'try',
		'500-error',
		array(
			'methods' => array( 'GET', 'POST' ),
			'callback' => static function () {
				return new WP_REST_Response( 'I am an error', 500 );
			},
			'permission_callback' => '__return_true',
		)
	);
} );

add_action(
	'wp_footer',
	static function () {
		?>
		<script type="module">
			const url = <?php echo wp_json_encode( rest_url( '/try/500-error/' ) ); ?>;
			fetch( url, { method: 'POST' } ).then( ( response ) => {
				console.info( response );
				return response.json();
			} ).then( ( json ) => {
				console.info( json );
			} ).catch( ( error ) => {
				console.warn( error );
			} );
		</script>
		<?php
	}
);
```

I am seeing a log entry coming from `console.warn()`:

> SyntaxError: Unexpected token '<', "<!DOCTYPE "... is not valid JSON

The service worker is getting the 500 error response back as expected:

![image](https://github.com/GoogleChromeLabs/pwa-wp/assets/134745/370ef5cd-4d92-4942-9ec2-5f65688d240a)

But then the service worker is incorrectly responding with the HTML error page:

![image](https://github.com/GoogleChromeLabs/pwa-wp/assets/134745/3b2cbc09-4bbf-4958-910b-d30c4f0cc677)

Interestingly, if I change the request method from `POST` to `GET` then this issue doesn't happen. It's only the `POST` request method that causes it. And this is due to the [`service-worker-offline-post-request-handling.js`](https://github.com/GoogleChromeLabs/pwa-wp/blob/56c64f107f2aa792f1d3d51056636f9347298f10/wp-includes/js/service-worker-offline-post-request-handling.js) logic which is matching all URLs:

https://github.com/GoogleChromeLabs/pwa-wp/blob/56c64f107f2aa792f1d3d51056636f9347298f10/wp-includes/js/service-worker-offline-post-request-handling.js#L147-L151

Clearly this should be updated to pass through the body's response if the response's `Content-Type` is JSON. Either this, or it should do nothing if it is not handling a navigation request (i.e. it is a fetch).

Here is some plugin code with a quick fix to disable this functionality:

```php
add_action(
	'wp_front_service_worker',
	static function ( WP_Service_Worker_Scripts $scripts  ) {
		unset( $scripts->registered['wp-offline-post-request-handling'] );
	},
	100
);
```

**Comment by reggieofarrell** [comment]
@westonruter that quick fix works for now. Thanks for your help.
